agent_app_sample_code/02_data_pipeline.py

```diff
@@ -0,0 +1,421 @@
+def file_parser(
+    raw_doc_contents_bytes: bytes,
+    doc_path: str,
+    modification_time: datetime,
+    doc_bytes_length: int,
+) -> ParserReturnValue:
+    """
+    Parses the content of a PDF document into a string.
+
+    This function takes the raw bytes of a PDF document and its path, attempts to parse the document using PyPDF,
+    and returns the parsed content and the status of the parsing operation.
+
+    Parameters:
+    - raw_doc_contents_bytes (bytes): The raw bytes of the document to be parsed (set by Spark when loading the file)
+    - doc_path (str): The path of the document, used to verify the file extension (set by Spark when loading the file)
+    - modification_time (timestamp): The last modification time of the document (set by Spark when loading the file)
+    - doc_bytes_length (long): The size of the document in bytes (set by Spark when loading the file)
+
+    Returns:
+    - ParserReturnValue: A dictionary containing the parsed document content and the status of the parsing operation.
+      The 'doc_content' key will contain the parsed text as a string, and the 'parser_status' key will indicate
+      whether the parsing was successful or if an error occurred.
+    """
+    try:
+        file_extension = doc_path.split(".")[-1]
+        parsed_document = {}
+
+        if file_extension == "pdf":
+            pdf = io.BytesIO(raw_doc_contents_bytes)
+            reader = PdfReader(pdf)
+
+            parsed_content = [
+                page_content.extract_text() for page_content in reader.pages
+            ]
+
+            parsed_document = {
+                "doc_content": "\n".join(parsed_content),
+                "parser_status": "SUCCESS",
+            }
+        elif file_extension == "html":
+            from markdownify import markdownify as md
+
+            html_content = raw_doc_contents_bytes.decode("utf-8")
+
+            markdown_contents = md(
+                str(html_content).strip(), heading_style=markdownify.ATX
+            )
+            markdown_stripped = re.sub(r"\n{3,}", "\n\n", markdown_contents.strip())
+
+            parsed_document = {
+                "doc_content": markdown_stripped,
+                "parser_status": "SUCCESS",
+            }
+        elif file_extension == "docx":
+            with tempfile.NamedTemporaryFile(delete=True) as temp_file:
+                temp_file.write(raw_doc_contents_bytes)
+                temp_file_path = temp_file.name
+                md = pypandoc.convert_file(temp_file_path, "markdown", format="docx")
+
+                parsed_document = {
+                    "doc_content": md.strip(),
+                    "parser_status": "SUCCESS",
+                }
+        else:
+            raise Exception(f"No supported parser for {doc_path}")
+
+        # Extract the required doc_uri
+        # convert from `dbfs:/Volumes/catalog/schema/pdf_docs/filename.pdf` to `Volumes/catalog/schema/pdf_docs/filename.pdf`
+        modified_path = doc_path[5:]
```

Comment:
This feels brittle, there is probably a better way? I think MLflow has some code for this

Reply:
replaced with `urlparse(doc_path).path.lstrip('/')` from urllib